Enforce required parameters for the various aws_s3 modes

willthames opened the pull request:
##### SUMMARY

Most modes require `object` parameter, and this is easy to
get wrong (e.g. through confusion with the `dest` parameter). As
it's as easy to enforce, let's do that.


##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
aws_s3

##### ANSIBLE VERSION
```
ansible 2.6.0 (devel e15a903bdf) last updated 2018/02/27 12:23:02 (GMT +1000)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/will/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/will/src/ansible/lib/ansible
  executable location = /home/will/src/ansible/bin/ansible
  python version = 2.7.14 (default, Jan 17 2018, 14:28:32) [GCC 7.2.1 20170915 (Red Hat 7.2.1-2)]
```

##### ADDITIONAL INFORMATION
No tests - as long as existing tests pass, I'm confident with this change (we don't need to check that Ansible does a good job of enforcing missing parameters here, I'd hope that's well tested elsewhere!)

gundalow commented:
+1 for adding more `required_if` checks

mattclay commented:
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/56225/67/tests

Commit:
Enforce required parameters for the various aws_s3 modes

Most modes require `object` parameter, and this is easy to
get wrong (e.g. through confusion with the `dest` parameter). As
it's as easy to enforce, let's do that.
